File: 300/396 - Rotate Function.c
#define max(a, b) ((a) > (b) ? (a) : (b))

int maxRotateFunction(int* A, int ASize) {
    int sum = 0, f = 0, i, ret;
    
    for (i = 0; i < ASize; ++i) {
        sum += A[i];
        f += i * A[i];
    }
    
    ret = f;
    for (i = 1; i < ASize; ++i) {
        f += sum - ASize * A[ASize - i];
        ret = max(f, ret);
    }
    
    return ret;
}

File: csharp/396.rotate-function.cs
/*
 * @lc app=leetcode id=396 lang=csharp
 *
 * [396] Rotate Function
 */
public class Solution {
    public int MaxRotateFunction(int[] A) {
        int n = A.Length;
        long ret, sum = 0, curr = 0;
        for (int i = 0; i < n; ++i) {
            curr += (long) i * A[i];
            sum += A[i];
        }
        ret = curr;
        for (int i = 0; i < n - 1; ++i) {
            curr += (long) n * A[i] - sum;
            ret = Math.Max(curr, ret);
        }
        return (int) ret;
    }
}

